### PR TITLE
OSOE-548: Upgrade to Orchard Core 1.6

### LINKED ISSUE #378: Upgrade to Orchard Core 1.6 (OSOE-548)
The new version is out, see https://docs.orchardcore.net/en/latest/docs/releases/1.6.0/ and https://github.com/OrchardCMS/OrchardCore/releases/tag/v1.6.0

- Utilize `AddIndexProvider` and `AddScopedIndexProvider` extensions, [see here](https://orcharddojo.net/blog/orchard-core-commerce-mvp-make-features-module-optional-to-non-default-tenant-this-week-in-orchard-03-02-2023#:~:text=the%20upcoming%20release.-,Add%20AddIndexProvider%20and%20AddScopedIndexProvider%20extensions,-This%20change%20is).

[Jira issue](https://lombiq.atlassian.net/browse/OSOE-548)

### PR DESCRIPTION
**Pull request by Psichorex**
Fixes #378 

[OSOE-548](https://lombiq.atlassian.net/browse/OSOE-548)

[OSOE-548]: https://lombiq.atlassian.net/browse/OSOE-548?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/Lombiq.OSOCE.Web/Program.cs [hunk 1]
```diff
@@ -19,7 +19,7 @@
         orchardCoreBuilder.ConfigureFeaturesGuard(
             new Dictionary<string, IEnumerable<string>>
             {
-                ["OrchardCore.Twitter"] = new[] { "Lombiq.UIKit", "Lombiq.ChartJs" },
+                ["OrchardCore.Twitter"] = new[] { "Lombiq.UIKit" },
```
Comment (Piedone) [inline src/Lombiq.OSOCE.Web/Program.cs]: You shouldn't change this. As I told under the issue: on `dev`, the UI test passes. If the UI test breaks with 1.6, then you should fix the underlying functionality, not employ a workaround like this.

File: pnpm-lock.yaml [hunk 2]
```diff
@@ -1,4 +1,4 @@
-lockfileVersion: 5.4
+lockfileVersion: 5.3
```
Comment (Piedone) [inline pnpm-lock.yaml]: Why the downgrade, and the other changes in this file necessary?

**Review by Piedone — CHANGES_REQUESTED** [review]

File: pnpm-lock.yaml (hunk 2 above)
Comment (Psichorex) [reply pnpm-lock.yaml]: It was probably unintentional as pnpm-lock.yaml is popping up constantly without me touching it.

File: src/Lombiq.OSOCE.Web/Program.cs (hunk 1 above)
Comment (Psichorex) [reply src/Lombiq.OSOCE.Web/Program.cs]: There is a problem related most probably to Orchard Core I am going to discuss now here in detail so you get a good understanding.

This is the code that the current dev contains:
image

Because the Conditinal Twitter module has 2 conditions the IFeatureEventHandler will run fully 2 times.
This is the first run :
image
Here you can see that the featuresToEnable contains Twitter only and it will call :
`await _shellFeatureManager.EnableFeaturesAsync(featuresToEnable)` 
OrchardCore.Twitter gets enabled here.

Problem roots here :
image
Here you see that the `shellDescriptor.Features` contains exactly 59 items.

After Twitter getting enabled and the IFeatureEventHandler runing the second time fully because it hit the second condition feature the shellDescriptor.Features will again contain 59 items. 

So basically it does not indicate that Twitter has been enabled.

So it tries to activate it again and this happens:
image

This was causing the UI Test problems but and this is an important but those test were flaky and they started passing after a certain time.
It was enough for me to take notice that something is fishy that's how I found out that this part is not functioning properly.

File: pnpm-lock.yaml (hunk 2 above)
Comment (0liver) [reply pnpm-lock.yaml]: Update to pnpm@7, please.

File: src/Lombiq.OSOCE.Web/Program.cs (hunk 1 above)
Comment (Psichorex) [reply src/Lombiq.OSOCE.Web/Program.cs]: Added a workaround that fixes the issue.

File: pnpm-lock.yaml (hunk 2 above)
Comment (Psichorex) [reply pnpm-lock.yaml]: @0liver It won't actually let me update PNPM. 
Giving npm install -g pnpm@7 will produce some installing like behavior but even after that checking pnpm -v give me back pnpm 6.32.0.

File: pnpm-lock.yaml [hunk 3]
```diff
@@ -1326,4 +1326,4 @@ packages:
   /yocto-queue/0.1.0:
     resolution: {integrity: sha512-rVksvsnNCdJ/ohGc6xgPwyN8eheCxsiLM8mxuE/t/mOVqJewPuO1miLpTHQiRgTKCLexL4MeAFVagts7HmNZ2Q==}
     engines: {node: '>=10'}
-    dev: true
```
Comment (Piedone) [inline pnpm-lock.yaml]: Revert the whitespace change too.

File: src/Lombiq.OSOCE.Web/Program.cs (hunk 1 above)
Comment (Piedone) [reply src/Lombiq.OSOCE.Web/Program.cs]: Thanks for taking the time to explain. However, three things to note:

1. Where do you see the test being flaky? Flaky means it sometimes fails, sometimes passes. I checked out the last 10 successful `dev` build (the most recent one was a timeout due to https://github.com/Lombiq/UI-Testing-Toolbox/issues/228) and none of them showed this test being flaky. I.e, it passed on the first time in each run. Or did it flakily pass after 1.6?
2. The goal of this issue is to upgrade the OSOCE solution to OC 1.6. If something used to work under 1.5 but then breaks 1.6 then that's something that needs to be fixed. Why it used to work, and whether the implementation was a good idea at that time might be interesting, but is ultimately beside the point; it needs to keep working. Finding incompatibilities and thus things breaking between OC versions is part and parcel of such an issue.
3. I don't know if you intentionally use the word "workaround" instead of something like "fix". If you're just using these interchangeably without a specific meaning, then no problem; otherwise, I'd like to clarify that in all such cases, we're seeking a permanent, proper fix to a given problem, not just a workaround (which is usually used to mean a temporary fix to get the system into a working order without actually fixing the underlying problem). It might be that for some reason it's infeasible to implement a good fix, and we need to be content with just a workaround, but this is not such a case.

**Review by Piedone — CHANGES_REQUESTED** [review]

**Review by Piedone — APPROVED** [review]

**Review by Piedone — APPROVED** [review]

**Review by Piedone — APPROVED** [review]

**Review by Piedone — APPROVED** [review]

**Review by Piedone — APPROVED** [review]

**Review by Piedone — APPROVED** [review]

**Review by Piedone — CHANGES_REQUESTED** [review]
1. Be sure to update **all** Lombiq package references in all projects while doing a release gradually: https://lombiq.atlassian.net/wiki/spaces/DEV/pages/786857987/Managing+releases+of+open-source+projects#An-example-of-releasing-OSOCE-projects A lot of the Helpful Libraries package references our outdated. Search for `PackageReference Include="Lombiq` in the solution, and you shouldn't see any non-alpha version for those packages that you released.
2. Also update the NuGetTest solution to the latest packages. Without that, the new packages are not tested.

**Review by Piedone — APPROVED** [review]

**Review by Piedone — APPROVED** [review]
Just marking the merge from `issue/OSOE-641`.

File: .github/workflows/build-and-test.yml [hunk 4]
```diff
@@ -20,6 +20,8 @@ jobs:
       ui-test-parallelism: 0
       build-create-binary-log: "true"
       blame-hang-timeout: "5m"
+      print-config-summary: "true"
```
Comment (Piedone) [inline .github/workflows/build-and-test.yml]: Remove these before setting the issue to Done.

**Review by Piedone — APPROVED** [review]

**Review by Piedone — APPROVED** [review]
